### PR TITLE
Add runtime failure tracking to polling inputs

### PR DESCRIPTION
**Pull request by mpfz0r**
Inputs that are polling their data, can experience intermittent failures. So far, those were only logged as errors in the server.log. There was no way to find these failures in the UI.

Introduce a new input IOState "Failing" which indicates those errors.

Add a new InputFailureRecorder to MessageInput.launch() which then can be passed to the Transports of various Inputs, and replace the LOG.errors certain calls.
The InputFailureRecorder can also be used to clear the failing input state, once the error is resolved.

Inputs that want to make use of the new InputFailureRecorder should inherit their Transports from ThrottleableTransport2, which has the new launch/doLaunch signature.

In addition, the SystemNotification for failing Inputs could not handle multiple Inputs at the time. To accomodate for this, we introduce an optional "key" for SystemNotifications. This allows us to differentiate between multiple Inputs.

### CODE REVIEW

**Review by thll — CHANGES_REQUESTED** [review]

**Comment by thll** [comment]
@mpfz0r, is there a companion PR for the Forwarder? Unless we make the changes to `NotificationService.java` backwards compatible, we are going to need one:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project graylog-forwarder: Compilation failure: Compilation failure:
[ERROR] /Volumes/CaseSensitive/git/forwarder/forwarder/src/main/java/org/graylog/forwarder/config/NoopNotificationService.java:[16,7] error: NoopNotificationService is not abstract and does not override abstract method destroyAllByTypeAndKey(Type,String) in NotificationService
[ERROR] /Volumes/CaseSensitive/git/forwarder/forwarder/src/main/java/org/graylog/forwarder/config/NoopNotificationService.java:[48,4] error: method does not override or implement a method from a supertype
```

**Comment by thll** [comment]
I've also noticed that in the Forwarder, we are not considering a `FAILING` input to be running, so that requires a change as well:

```
2023-03-08T12:00:47.625+0100 INFO  [scheduled-daemon-9] control.configuration.ConfigurationProcessor - Server requested input [[JSON path/64086953975fbb4ef727ffaf]] to be started. Current state of input: [FAILING].
```

File: graylog2-server/src/main/java/org/graylog2/plugin/inputs/transports/ThrottleableTransport2.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        doLaunch(input, inputFailureRecorder);
+
+        // only listen for updates if we are allowed to be throttled at all
+        if (throttlingAllowed) {
+            eventBus.register(this);
+        }
+    }
+
+    @Override
+    protected void doLaunch(MessageInput input) throws MisfireException {
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/plugin/inputs/transports/ThrottleableTransport2.java:47]: ah. good idea.

File: graylog2-server/src/main/java/org/graylog2/plugin/InputFailureRecorder.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    public void isFailing(Class clazz, String error, @Nullable Throwable e) {
+        if (inputState.getState().equals(IOState.Type.FAILING)) {
+            return;
+        }
+        if (e != null) {
+            inputState.setState(IOState.Type.FAILING, error + ": (" + e.getMessage() + ")");
+        } else {
+            inputState.setState(IOState.Type.FAILING, error);
+        }
+        LoggerFactory.getLogger(clazz).warn(error, e);
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/plugin/InputFailureRecorder.java:46]: yeah, I'm gonna go for option 2 :+1:

File: graylog2-server/src/main/java/org/graylog2/plugin/InputFailureRecorder.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.plugin;
+
+import org.graylog2.plugin.inputs.MessageInput;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+public class InputFailureRecorder {
```
[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/plugin/InputFailureRecorder.java:24]: yeah. fixing

**Comment by mpfz0r** [comment]
> I've also noticed that in the Forwarder, we are not considering a `FAILING` input to be running, so that requires a change as well:

Thanks! 
This should do it: 
https://github.com/Graylog2/forwarder/pull/96
https://github.com/Graylog2/graylog-plugin-enterprise/pull/4822

**Review by thll — APPROVED** [review]

**Comment by thll** [comment]
Great feature. Thank you for going through with it :pray:

**Pull request merged**